### ISSUE TITLE
Improve athena research speed

### ISSUE THREAD

**Issue by Limezy** [athensresearch/athens#969]
**Problem**
I've recently started to import big text corpus into Athens (the whole Bible 😂)
This makes my bloc quotes research slow, because there are so much blocs to search through : doesn't bother me and it's normal (even though I hope it'll be improved over time)
This makes the opening time of a few pages that have big numbers of blocs slow : doesn't bother me (same remarks)

However, **my athena research bar is and feels now very slow, and that bothers me !**
Indeed, it looks like the Athena bar is currently : 
1. Registering what was typed (but not showing it)
2. Crawling through all pages AND all blocs
3. Showing its results

It seems to me that the research is mainly supposed to search through **pages**. It's nice to have some blocs proposals, but this should stay only a side thing. So instead of crawling through all blocs before showing the answer, I believe that the search bar should : 
1. Register AND show what was typed (to avoid the feeling of typing nothing)
2. Crawl through all pages
2. Show a first answer including only pages
3. In the background, crawl through all blocs
4. Show additional proposals based on blocs if the user is still on the bar

That would greatly improve the speed feeling !

**Athens Version**
1.0.0-BETA.68 - Intel OSX

**Comment by tangjeff0** [comment]
This is probably my biggest problem with Athens right now. Any ideas on how to improve query speed @neotyk? Right now we index through ALL blocks AND pages. Pages would be a lot faster, but we want a way to search through all blocks as well.

**Comment by neotyk** [comment]
After a quick research, here are my 5 cents:

1. Debounce Athena's input, once Socrates stops typing we issue a search query.
2. Make query start and query results asynchronous, currently every keystroke performs search.
3. Communicate that we're busy with search.

We could perform 3 searches with WebWorkers in parallel, after all why all those cores, right?

**Comment by sawhney17** [comment]
Perhaps, on the front end, something similar to the way alfred handles everything. It by default only searches apps, although you can change the default search scope in settings and then you click space or type a apostrophe prior to the query and then it does a full search.

**Comment by Limezy** [comment]
I'm not a programmer, but what about making cache with the most common researches ?

**Comment by tangjeff0** [comment]
> Debounce Athena's input, once Socrates stops typing we issue a search query.

Going to do this first @neotyk.

> Perhaps, on the front end, something similar to the way alfred handles everything. It by default only searches apps, although you can change the default search scope in settings and then you click space or type a apostrophe prior to the query and then it does a full search.

@sawhney17 this is a crazy good idea. We are also starting to think about running executing commands from Athena. There are a few ways to distinguish "search" and "run", and would love to hear everyone's thoughts on this (+1 @shanberg).

Alfred: `'` scope search to contents

Intelli-j: separate keybindings for different kinds of search. Demo: https://www.loom.com/share/43a4312ba191465c93d1ab9fd27f977b

Superhuman

**Comment by sboesen** [comment]
Definitely second discord style "prefixed" search, I use it pretty regularly and it works well. Part of what I like about their implementation is it always shows the list of prefixes when you click in to search, so there's no "how do I search in url on google again?":
<img width="322" alt="Screen Shot 2021-04-16 at 11 21 07 AM" src="https://user-images.githubusercontent.com/248703/115067564-f6a91280-9ea5-11eb-87a7-d528a7355d24.png">

Something like "block:" or maybe "b:" seems sensible to me.

**Comment by Limezy** [comment]
Hi, thank you so much for pulling out that .69 version so quick.

Now the Athena search bar has become really usable again despite my ~32k blocks in ~80 pages 💪

However there is something I really don't understand !
* When I search for a random block in the Athena search bar, it takes roughly 2 seconds to show up.
* When I search **for the exact same random block** as a block quote - after typing (( in a page, it takes 25 seconds to show up !!!

Maybe I have misunderstood about something going on, but for me the software has to go through the same search tree, no ?
How such a difference can be explained ?

If the block quote could be as fast as the Athena search then I'll definitely stack all my content in Athens 👍

**Comment by tangjeff0** [comment]
Hey @Limezy, I tried to implement the delayed search for block references, but the implementation made it difficult. Refactoring keyboard input is a priority, which would address this. Block refs are actually kinda unusable for me because they are so slow on my db... :(

**Comment by Limezy** [comment]
I see. But anyway now I know thanks to Athena how quick the search tree can be, despite my very high number of blocks.
So I can hope that one day my block quotes will be as quick (if not quicker when hitchhiker trees will arrive).
I'm already quite pleased with the performance, given how old is my main Mac (2006 - 15 years old buddy)

**Comment by neotyk** [comment]
New parser release should improve performance on heavy pages. Unfortunately it's not addressing search times.

**Comment by tangjeff0** [comment]
Closed by #756
